phoenix-core/src/main/java/org/apache/phoenix/index/GlobalIndexChecker.java

```diff
@@ -0,0 +1,366 @@
+                    // skip this row as it is invalid
+                    // if there is no more row, then result will be an empty list
+                } while (hasMore);
+                return hasMore;
+            } catch (Throwable t) {
+                ServerUtil.throwIOException(region.getRegionInfo().getRegionNameAsString(), t);
+                return false; // impossible
+            }
+        }
+
+        private void deleteRowIfAgedEnough(byte[] indexRowKey, long ts) throws IOException {
+            if ((EnvironmentEdgeManager.currentTimeMillis() - ts) > ageThreshold) {
+                Delete del = new Delete(indexRowKey);
+                Mutation[] mutations = new Mutation[]{del};
+                region.batchMutate(mutations);
+            }
+        }
+
+        private List<Cell> rebuildSingleIndexRow(byte[] indexRowKey, long ts, List<Cell> row) throws IOException {
+            // Build the data table row key from the index table row key
+            if (buildIndexScan == null) {
+                buildIndexScan = new Scan();
+                indexScan = new Scan(scan);
+                byte[] dataTableName = scan.getAttribute(PHYSICAL_DATA_TABLE_NAME);
+                byte[] indexTableName = region.getRegionInfo().getTable().getName();
+                dataHTable = ServerUtil.getHTableForCoprocessorScan(env,
+                        SchemaUtil.getPhysicalTableName(dataTableName, env.getConfiguration()));
+                if (indexMaintainer == null) {
+                    byte[] md = scan.getAttribute(PhoenixIndexCodec.INDEX_PROTO_MD);
+                    List<IndexMaintainer> maintainers = IndexMaintainer.deserialize(md, true);
+                    indexMaintainer = getIndexMaintainer(maintainers, indexTableName);
+                }
+                if (indexMaintainer == null) {
+                    throw new DoNotRetryIOException(
+                            "rebuildSingleIndexRow: IndexMaintainer is not included in scan attributes for " +
+                                    region.getRegionInfo().getTable().getNameAsString());
+                }
+                if (viewConstants == null) {
+                    viewConstants = IndexUtil.deserializeViewConstantsFromScan(scan);
+                }
+                buildIndexScan.setAttribute(BaseScannerRegionObserver.UNGROUPED_AGG, TRUE_BYTES);
+                buildIndexScan.setAttribute(PhoenixIndexCodec.INDEX_PROTO_MD, scan.getAttribute(PhoenixIndexCodec.INDEX_PROTO_MD));
+                buildIndexScan.setAttribute(BaseScannerRegionObserver.REBUILD_INDEXES, TRUE_BYTES);
+            }
+            // Rebuild the index row from the corresponding the row in the the data table
+            byte[] dataRowKey = indexMaintainer.buildDataRowKey(new ImmutableBytesWritable(indexRowKey), viewConstants);
+            buildIndexScan.withStartRow(dataRowKey, true);
+            buildIndexScan.withStopRow(dataRowKey, true);
+            buildIndexScan.setRaw(true);
```

Comment:
Why do we rebuild from the data table with a raw scan? Wouldn't we want to know if data's been deleted or expired out?

Reply:
I am following the partial rebuild code here. If the data table row is deleted, then the index row should be deleted too.  In order to get the delete marker from the data table, we need to do a raw scan on it.